File: algorithms/c++/21-merge-two-sorted-lists.cpp
// https://leetcode.com/problems/merge-two-sorted-lists/
#include <bits/stdc++.h>

using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode *list = new ListNode;
        if (list1 == nullptr) { return list2; }
        if (list2 == nullptr) { return list1; }
        if (list1->val <= list2->val) {
            list->val = list1->val;
            list->next = mergeTwoLists(list1->next, list2);
        } else {
            list->val = list2->val;
            list->next = mergeTwoLists(list1, list2->next);
        }
        return list;
    }
};

class Solution2 {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* front = new ListNode();
        ListNode* curr = front;
        
        while(list1 && list2) {
            if (list1->val <= list2->val) {
                curr->next = list1;
                list1 = list1->next;
            } else {
                curr->next = list2;
                list2 = list2->next;
            }
            curr = curr->next;
        }
        
        if (list1)
            curr->next = list1;
        if (list2)
            curr->next = list2;
        
        return front->next;
    }
};

int main() {
    
    return 0;
}

File: algorithms/swift/21-merge_two_sorted_lists.swift
/*
Link: https://leetcode.com/problems/merge-two-sorted-lists/

*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func mergeTwoLists(_ list1: ListNode?, _ list2: ListNode?) -> ListNode? {
        if list1 == nil && list2 == nil { return nil }
        else if list1 == nil && list2 != nil { return list2 }
        else if list2 == nil && list1 != nil { return list1 }
        
        var head = ListNode()
        
        var prev: ListNode? = nil
        var a: ListNode? = list1
        var b: ListNode? = list2
        if a!.val <= b!.val {
            head.val = a!.val
            a = a!.next
        } else {
            head.val = b!.val
            b = b!.next
        }
        prev = head
        while (a != nil && b != nil) {
            if a!.val <= b!.val {
                prev!.next = ListNode(a!.val)
                a = a!.next
            } else {
                prev!.next = ListNode(b!.val)
                b = b!.next
            }
            prev = prev!.next
        }
        if (a == nil && b != nil) {
            prev!.next = b
        } else if (a != nil && b == nil) {
            prev!.next = a
        }
        
        return head
    }
}